politeiad/api/v2/v2.go

```diff
@@ -0,0 +1,512 @@
+// Copyright (c) 2020-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package v2
+
+import "fmt"
+
+const (
+	// APIRoute is prefixed onto all routes in this package.
+	APIRoute = "/v2"
+
+	// Routes
+	RouteRecordNew           = "/recordnew"
+	RouteRecordEdit          = "/recordedit"
+	RouteRecordEditMetadata  = "/recordeditmetadata"
+	RouteRecordSetStatus     = "/recordsetstatus"
+	RouteRecordGet           = "/recordget"
+	RouteRecordGetBatch      = "/recordgetbatch"
+	RouteRecordGetTimestamps = "/recordgettimestamps"
+	RouteInventory           = "/inventory"
+	RoutePluginWrite         = "/pluginwrite"
+	RoutePluginReads         = "/pluginreads"
+	RoutePluginInventory     = "/plugininventory"
+
+	// ChallengeSize is the size of a request challenge token in bytes.
+	ChallengeSize = 32
+)
+
+// ErrorCodeT represents a user error code.
+type ErrorCodeT uint32
+
+const (
+	ErrorCodeInvalid                 ErrorCodeT = 0
+	ErrorCodeRequestPayloadInvalid   ErrorCodeT = 1
+	ErrorCodeChallengeInvalid        ErrorCodeT = 2
+	ErrorCodeMetadataStreamInvalid   ErrorCodeT = 3
+	ErrorCodeMetadataStreamDuplicate ErrorCodeT = 4
+	ErrorCodeFilesEmpty              ErrorCodeT = 5
+	ErrorCodeFileNameInvalid         ErrorCodeT = 6
+	ErrorCodeFileNameDuplicate       ErrorCodeT = 7
+	ErrorCodeFileDigestInvalid       ErrorCodeT = 8
+	ErrorCodeFilePayloadInvalid      ErrorCodeT = 9
+	ErrorCodeFileMIMETypeInvalid     ErrorCodeT = 10
+	ErrorCodeFileMIMETypeUnsupported ErrorCodeT = 11
+	ErrorCodeTokenInvalid            ErrorCodeT = 12
+	ErrorCodeRecordNotFound          ErrorCodeT = 13
+	ErrorCodeRecordLocked            ErrorCodeT = 14
+	ErrorCodeNoRecordChanges         ErrorCodeT = 15
+	ErrorCodeStatusChangeInvalid     ErrorCodeT = 16
+	ErrorCodePluginIDInvalid         ErrorCodeT = 17
+	ErrorCodePluginCmdInvalid        ErrorCodeT = 18
+)
+
+var (
```

Comment:
We should have a test that counts the number of errors and compares it with the map. That way we can detect if someone adds a new error but fails to add the string representation.

Reply:
How do you count the numbers of errors without hard coding it?  In the example above, how do you know that errors 0-18 are being used and the length of the map should be 19.

Reply:
You add a `ErrorCodeLast` which is always the last entry in the constants. Then you add a test that checks the `ErrorCodeLast` against the length of the map. If they aren't identical then someone forgot to add the human readable error. Simple trick but has saved me countless times in dcrd, for example.

Reply:
Cool, I'll add that everywhere.